### ISSUE TITLE
[request] Encode/decode variables

### ISSUE THREAD

**Issue by stsievert** [pytorch/pytorch#2787]
I'd like to encode/decode the gradient. This will require storing arbitrary values in memory instead of the gradient.

That is, I'd like to perform

``` python
def encode(name):
    def hook(var):
        return {'name': name, 'grad': var.grad}
    return hook
def decode(hook_output):
    return hook_output['grad']

for name, param in model.named_paramers():
    param.register_hook(encode(name))

# later; psuedo-code
for hook_out in hook_outputs:
    grad = decode(hook_output)
```

but this throws an error at [python_hook.cpp#L147].

[python_hook.cpp#L147]:https://github.com/pytorch/pytorch/blob/1290e586fbc3d6266423f3417723d6620267054b/torch/csrc/autograd/python_hook.cpp#L147

**Comment by stsievert** [comment]
I can see two solutions to this issue:

1. Having `encode` and `decode` functions on class Variable. Every time the variable is computed/stored these functions will be called.
2. Allowing `register_hook` to accept a keyword argument `check`, which will default to `True`. if `check == True`, [python_hook.cpp#L147] will be run.

[python_hook.cpp#L147]:https://github.com/pytorch/pytorch/blob/1290e586fbc3d6266423f3417723d6620267054b/torch/csrc/autograd/python_hook.cpp#L147

**Comment by albanD** [comment]
I am not sure to understand what you want to do here.
You have a `FuncA` that outputs a `Variable` of size `10x10`, this is then used in `FuncB` for further processing.
You want to put a hook on the Variable to change it's gradient to something else than a `10x10` Tensor/Variable.
But when you keep doing backpropagation, `FuncA` is expecting a Tensor/Variable of size `10x10` for its output, it won't be able to process your modified object !

What is it exactly that you're trying to achieve?

**Comment by stsievert** [comment]
Sorry for the explanation.

What I'm trying to do is similar to the [QSGD paper]. It approximates the gradient by reducing the gradient precision (from 64 bit floats to 8 bits). My use case would rather want to store a dict, or at least a tensor of a different size.

This is a coding scheme that can both encode and decode the gradient. I'd like the `encode` function to be called every time the gradient is computed and overwrite the gradient, and `decode` to be called every time the gradient is used. Something like

``` python
def encode(var):
    return {'grad': var.grad, 'other': stuff}
def decode(encode_output):
    return encode_output['grad']

# in setup
for param in model.named_parameters():
    param.encoding(encode)
    param.decoding(decode)
```

Does that make more sense?

[QSGD paper]:https://arxiv.org/abs/1610.02132

**Comment by stsievert** [comment]
I want to resolve this issue for one of my classes final project. Can one of you look over [my proposal][1]? I'd appreciate comments on issues I might face or where to start developing.

[1]:https://www.dropbox.com/sh/7uryl07t2tkkhu9/AABgZcBgJhlPXkbXSV0_gFkga?dl=0

**Comment by apaszke** [comment]
I still can't see how would we implement this. It would require us to constantly encode and decode the gradients, because autograd functions are implemented in such a way that they can only work with floating point gradients contained in Variables. I'm not entirely sure what's your use case? Do you want to quantize the intermediate gradients (so sort of have a hook that is called for each intermediate grad)?

**Comment by stsievert** [comment]
>  I'm not entirely sure what's your use case?

My use case mirrors the one in [the QSGD paper][qsgd]. That is, I would like to encode/decode gradients to preserve bandwidth on the CPU–GPU connection to avoid communication bottlenecks.

Bandwidth is limited is limited between GPU–GPU and CPU–GPU, leading to communication bottlenecks. I'd like to perform something like below to avoid this:

<img width="879" alt="screen shot 2017-10-17 at 10 59 32 am" src="https://user-images.githubusercontent.com/1320475/31675449-4b24c6c6-b32a-11e7-99bb-c756c653370a.png">

All blocks exist in PyTorch right now except the encode/decode blocks, and I don't want to touch them. For my use case device1 and device2 are on the same machine (and it looks like they're both GPUs in the QSGD paper).

> Do you want to quantize the intermediate gradients (so sort of have a hook that is called for each intermediate grad)?

No, I do not want to quantize each intermediate gradient. I want a forward hook that is only called after computation of the gradient is finished.

[qsgd]:https://arxiv.org/abs/1610.02132

**Comment by stsievert** [comment]
There's no reason I can't make a custom optimizer to perform this task.

I'll also experiment with removing the check https://github.com/pytorch/pytorch/blob/1290e586fbc3d6266423f3417723d6620267054b/torch/csrc/autograd/python_hook.cpp#L147

**Comment by mruberry** [comment]
Closing due to age. If this is still an issue please comment and reopen it.
